tket2/src/serialize/pytket/param/encode.rs

```diff
@@ -0,0 +1,130 @@
+
+use hugr::ops::{OpType, Value};
+use hugr::std_extensions::arithmetic::float_ops::FloatOps;
+use hugr::std_extensions::arithmetic::float_types::ConstF64;
+
+use crate::extension::rotation::{ConstRotation, RotationOp};
+use crate::extension::sympy::SympyOp;
+use crate::ops::match_symb_const_op;
+
+/// Fold a rotation or float operation into a string, given the string
+/// representations of its inputs.
+///
+/// The folded op must have a single string output.
+///
+/// Returns `None` if the operation cannot be folded.
+pub fn fold_param_op(optype: &OpType, inputs: &[&str]) -> Option<String> {
+    let param = match optype {
+        OpType::Const(const_op) => {
+            // New constant, register it if it can be interpreted as a parameter.
+            match try_constant_to_param(const_op.value()) {
+                Some(param) => param,
+                None => return None,
+            }
+        }
+        OpType::LoadConstant(_op_type) => {
+            // Re-use the parameter from the input.
+            inputs[0].to_string()
+        }
+        // Encode some angle and float operations directly as strings using
+        // the already encoded inputs. Fail if the operation is not
+        // supported, and let the operation encoding process it instead.
+        OpType::ExtensionOp(_) => {
+            if let Some(s) = optype
+                .cast::<RotationOp>()
+                .and_then(|op| encode_rotation_op(&op, inputs))
+            {
+                s
+            } else if let Some(s) = optype
+                .cast::<FloatOps>()
+                .and_then(|op| encode_float_op(&op, inputs))
+            {
+                s
+            } else if let Some(s) = optype
+                .cast::<SympyOp>()
+                .and_then(|op| encode_sympy_op(&op, inputs))
+            {
+                s
+            } else {
+                return None;
+            }
+        }
+        _ => match_symb_const_op(optype)?.to_string(),
+    };
+    Some(param)
+}
+
+/// Convert a HUGR rotation or float constant to a TKET1 parameter.
+///
+/// Angle parameters in TKET1 are encoded as a number of half-turns,
+/// whereas HUGR uses radians.
+#[inline]
+fn try_constant_to_param(val: &Value) -> Option<String> {
+    if let Some(const_angle) = val.get_custom_value::<ConstRotation>() {
+        // Special case for pi rotations
+        if const_angle == &ConstRotation::PI {
+            return Some("pi".to_string());
+        } else if const_angle == &ConstRotation::PI_2 {
+            return Some("pi/2".to_string());
+        } else if const_angle == &ConstRotation::PI_4 {
+            return Some("pi/4".to_string());
+        } else if const_angle == &ConstRotation::TAU {
+            return Some("2pi".to_string());
+        }
+
+        let half_turns = const_angle.half_turns();
+        Some(half_turns.to_string())
+    } else if let Some(const_float) = val.get_custom_value::<ConstF64>() {
```

Comment:
pytket param expressions are also in half_turns, so no need to convert here.

The problem actually was in the decoding of `pi` constants. That should be decoded to the `3.1415`... float constant rather than a rotation with value `1`.
This lets us write expressions like `2alpha / pi` correctly.